File: src/algorithms/arrays/sorting-partitioning/lomuto-partition/__tests__/lomuto-partition_test.go
package lomutopartition

import (
	"reflect"
	"testing"
)

func TestDefaultInputPivotAtCorrectPosition(t *testing.T) {
	pivotIndex, result := lomutoPartition([]int{8, 3, 6, 1, 5, 9, 2, 7})
	if result[pivotIndex] != 7 {
		t.Errorf("Expected pivot value 7 at pivotIndex=%d, got %d", pivotIndex, result[pivotIndex])
	}
	for _, leftVal := range result[:pivotIndex] {
		if leftVal > 7 {
			t.Errorf("Left element %d > pivot 7", leftVal)
		}
	}
	for _, rightVal := range result[pivotIndex+1:] {
		if rightVal <= 7 {
			t.Errorf("Right element %d should be > pivot 7", rightVal)
		}
	}
}

func TestSortedArrayPivotAtLast(t *testing.T) {
	pivotIndex, result := lomutoPartition([]int{1, 2, 3, 4, 5})
	if pivotIndex != 4 {
		t.Errorf("Expected pivotIndex=4, got %d", pivotIndex)
	}
	if result[4] != 5 {
		t.Errorf("Expected result[4]=5, got %d", result[4])
	}
}

func TestReverseSortedPivotAtFirst(t *testing.T) {
	pivotIndex, result := lomutoPartition([]int{5, 4, 3, 2, 1})
	if pivotIndex != 0 {
		t.Errorf("Expected pivotIndex=0, got %d", pivotIndex)
	}
	if result[0] != 1 {
		t.Errorf("Expected result[0]=1, got %d", result[0])
	}
}

func TestSingleElement(t *testing.T) {
	pivotIndex, result := lomutoPartition([]int{42})
	if pivotIndex != 0 {
		t.Errorf("Expected pivotIndex=0, got %d", pivotIndex)
	}
	if !reflect.DeepEqual(result, []int{42}) {
		t.Errorf("Expected [42], got %v", result)
	}
}

func TestEmptyArray(t *testing.T) {
	pivotIndex, result := lomutoPartition([]int{})
	if pivotIndex != -1 {
		t.Errorf("Expected pivotIndex=-1 for empty, got %d", pivotIndex)
	}
	if len(result) != 0 {
		t.Errorf("Expected empty result, got %v", result)
	}
}

func TestTwoElementsLargerFirst(t *testing.T) {
	pivotIndex, result := lomutoPartition([]int{5, 2})
	if pivotIndex != 0 {
		t.Errorf("Expected pivotIndex=0, got %d", pivotIndex)
	}
	if result[0] != 2 || result[1] != 5 {
		t.Errorf("Expected [2,5], got %v", result)
	}
}

File: src/algorithms/arrays/sorting-partitioning/lomuto-partition/__tests__/LomutoPartition_test.java
import java.util.Arrays;

public class LomutoPartition_test {
    public static void main(String[] args) {
        // Default input: pivot 7 at correct position, all left <= 7, all right > 7
        {
            int[] result = LomutoPartition.lomutoPartition(new int[]{8, 3, 6, 1, 5, 9, 2, 7});
            // result[0] = pivotIndex, result[1..] = partitioned array
            int pivotIndex = result[0];
            assert result[pivotIndex + 1] == 7 : "Pivot value should be 7";
            for (int leftIdx = 1; leftIdx < pivotIndex + 1; leftIdx++) {
                assert result[leftIdx] <= 7 : "Left element should be <= 7";
            }
            for (int rightIdx = pivotIndex + 2; rightIdx < result.length; rightIdx++) {
                assert result[rightIdx] > 7 : "Right element should be > 7";
            }
        }

        // Already sorted [1,2,3,4,5]: pivot 5 at last index
        {
            int[] result = LomutoPartition.lomutoPartition(new int[]{1, 2, 3, 4, 5});
            assert result[0] == 4 : "Expected pivotIndex=4, got " + result[0];
            assert result[5] == 5 : "Expected pivot value=5, got " + result[5];
        }

        // Reverse sorted [5,4,3,2,1]: pivot 1 at index 0
        {
            int[] result = LomutoPartition.lomutoPartition(new int[]{5, 4, 3, 2, 1});
            assert result[0] == 0 : "Expected pivotIndex=0, got " + result[0];
            assert result[1] == 1 : "Expected pivot value=1, got " + result[1];
        }

        // Single element [42]
        {
            int[] result = LomutoPartition.lomutoPartition(new int[]{42});
            assert result[0] == 0 : "Expected pivotIndex=0, got " + result[0];
            assert result[1] == 42 : "Expected value=42, got " + result[1];
        }

        // Empty array
        {
            int[] result = LomutoPartition.lomutoPartition(new int[]{});
            assert result[0] == -1 : "Expected pivotIndex=-1 for empty, got " + result[0];
        }

        // Two elements [5,2]: pivot 2 goes to index 0
        {
            int[] result = LomutoPartition.lomutoPartition(new int[]{5, 2});
            assert result[0] == 0 : "Expected pivotIndex=0, got " + result[0];
            assert result[1] == 2 : "Expected result[0]=2, got " + result[1];
            assert result[2] == 5 : "Expected result[1]=5, got " + result[2];
        }

        System.out.println("All tests passed!");
    }
}
